src/vidpf.rs

```diff
@@ -0,0 +1,735 @@
+            y = self.eval_next(key.id, alpha_i, &mut s_i, &mut t_i, &pk.cw[i]);
+            pi = self.proof_next(&pi, alpha, i, &s_i, t_i, &pk.cs[i]);
+        }
+
+        Share { y, pi }
+    }
+
+    /// verify checks whether the proofs are equal and that the shares add up to beta.
+    pub fn verify(&self, a: &Share<R>, b: &Share<R>, beta: &R) -> bool {
+        assert!(a.pi.0.len() == self.proof_size);
+        assert!(b.pi.0.len() == self.proof_size);
+        &a.y + &b.y == *beta && a.pi.0 == b.pi.0
+    }
+
+    /// initial_proof returns the initial value of a Proof for the root level.
+    pub fn initial_proof(&self) -> Proof {
+        Proof::new(self.proof_size)
+    }
+
+    /// initial_seed returns the initial value of a Seed for the root level.
+    pub fn initial_seed(&self) -> Seed {
+        Seed::new(self.seed_size)
+    }
+
+    fn prg(&self, seed: &Seed) -> Sequence {
+        let dst = "100".as_bytes();
+        let mut prg = self.prng_ctor.new_prng(seed, dst, &self.bind_info);
+
+        let mut sl = self.initial_seed();
+        let mut sr = self.initial_seed();
+        let mut tl = ControlBit::from(0);
+        let mut tr = ControlBit::from(0);
+        sl.fill(&mut prg);
+        sr.fill(&mut prg);
+        tl.fill(&mut prg);
+        tr.fill(&mut prg);
+
+        Sequence(sl, tl, sr, tr)
+    }
+
+    fn convert(&self, seed: &Seed) -> (Seed, R) {
+        let dst = "101".as_bytes();
+        let mut prg = self.prng_ctor.new_prng(seed, dst, &self.bind_info);
+
+        let mut out_seed = self.initial_seed();
+        let mut value = R::new();
+        out_seed.fill(&mut prg);
+        value.fill(&mut prg);
+
+        (out_seed, value)
+    }
+
+    fn hash_one(&self, alpha: &[u8], level: usize, seed: &Seed) -> Proof {
+        let dst = "vidpf cs proof".as_bytes();
+        let mut binder = Vec::new();
+        binder.extend_from_slice(alpha);
+        binder.extend(level.to_le_bytes());
+        let mut prg = self.prng_ctor.new_prng(seed, dst, &binder);
+
+        let mut proof = self.initial_proof();
+        proof.fill(&mut prg);
+
+        proof
+    }
+
+    fn hash_two(&self, proof: &Proof) -> Proof {
+        let dst = "vidpf proof adjustment".as_bytes();
+        let seed = self.initial_seed();
+        let binder = &proof.0;
+        let mut prg = self.prng_ctor.new_prng(&seed, dst, binder);
+
+        let mut out_proof = self.initial_proof();
+        out_proof.fill(&mut prg);
+
+        out_proof
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+/// ServerID used to identify two aggregation servers.
+pub enum ServerID {
+    /// S0 is the first server.
+    S0 = 0,
+    /// S1 is the second server.
+    S1 = 1,
+}
+
+#[derive(Debug)]
+/// Share
+pub struct Share<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    /// y is the sharing of the output.
+    pub y: R,
+    /// pi is a proof used to verify the share.
+    pub pi: Proof,
+}
+
+/// RangeOps
```

Comment:
What is the puprose of `RangeOps`, `Range`, and `Fill`? Are these redefined from the standard library? Probably we don't actually need to re-define them here.

Reply:
Range defines the codomain of the PDF function F:{0,1}*-->Range.
Fill was removed now.